File: 백준/Silver/2343. 기타 레슨/기타 레슨.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] lectures = new int[N];
        int start = 0; // 최소 블루레이 크기
        int end = 0; // 최대 블루레이 크기

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            lectures[i] = Integer.parseInt(st.nextToken());

            start = Math.max(start, lectures[i]); // 최소 크기는 가장 큰 강의 길이
            end += lectures[i]; // 최대 크기는 모든 강의 길이의 합
        }

        int result = end; // 결과값 저장 (최소 블루레이 크기)

        while (start <= end) {
            int mid = (start + end) / 2; // 블루레이 크기의 중간값
            int sum = 0;
            int cnt = 1; // 블루레이 개수 (최소 1개)

            for (int i = 0; i < N; i++) {
                if (sum + lectures[i] > mid) {
                    cnt++;
                    sum = lectures[i];
                } else {
                    sum += lectures[i];
                }
            }

            if (cnt <= M) { // 블루레이 개수가 적거나 같으면 크기 줄이기
                result = mid;
                end = mid - 1;
            } else { // 블루레이 개수가 많으면 크기 늘리기
                start = mid + 1;
            }
        }

        System.out.println(result); // 최소 블루레이 크기 출력
    }
}

File: 백준/Silver/2343. 기타 레슨/기타 레슨.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

int main()
{
	std::ios::sync_with_stdio(false);
	std::cin.tie(nullptr);

	int N, M;
	std::cin >> N >> M;

	std::vector<int> lectures(N);
	int max = 0;
	int min = 0;

	for (int i = 0; i < N; i++) {
		std::cin >> lectures[i];
		min = std::max(min, lectures[i]);
		max += lectures[i];
	}

	int result = 0;

	while (min <= max)
	{
		int mid = (min + max) / 2;
		int sum = 0;
		int cnt = 1;

		for (int i = 0;i < N; i++) {
			if (lectures[i] + sum > mid)
			{
				cnt++;
				sum = lectures[i];
			}
			else
			{
				sum += lectures[i];
			}
		}

		if (cnt <= M)
		{
			result = mid;
			max = mid - 1;
		} 
		else
		{
			min = mid + 1;
		}
	}

	std::cout << result;

	return 0;
}